### PR TITLE
Add an entitiesHash to Keep

### PR DESCRIPTION
**Pull request by ryanpbrewster**
The entitiesHash encodes the set of (libraries, users) that are connected
to a Keep. This will allow us to quickly find Keeps that have exactly
the same set of connected entities.

I had a really obnoxious bug where I was serializing the Keep into Json
incorrectly, so the Keep cache was getting garbage data (but the DB was fine).
When keepRepo.get was called it was automatically looking in the cache and
pulling out the garbage data. Tough to track down.

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
@leogrim @andrewconner @stkem

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 1]
```diff
@@ -180,12 +191,33 @@ object Keep {
         "sourceAttributionId" -> k.sourceAttributionId,
         "note" -> k.note,
         "originalKeeperId" -> k.originalKeeperId.orElse(Some(k.userId)),
-        "organizationId" -> k.organizationId
+        "organizationId" -> k.organizationId,
+        "entitiesHash" -> k.entitiesHash
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: @andrewconner I forgot this line

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 2]
```diff
@@ -31,9 +33,10 @@ case class Keep(
     sourceAttributionId: Option[Id[KeepSourceAttribution]] = None,
     note: Option[String] = None,
     originalKeeperId: Option[Id[User]] = None,
-    organizationId: Option[Id[Organization]] = None) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
+    organizationId: Option[Id[Organization]] = None,
+    entitiesHash: Option[EntitiesHash] = None) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
 
-  def sanitizeForDelete: Keep = copy(title = None, state = KeepStates.INACTIVE)
+  def sanitizeForDelete: Keep = copy(title = None, state = KeepStates.INACTIVE, isPrimary = false, entitiesHash = Some(KeepConnectionEntities.empty.hash)) // good idea?
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: @leogrim this is new and relates to things we've talked about in person. Dead keeps should have `isPrimary = false`. As soon as we drop the constraint on 1 URI / library I'm going to kill `isPrimary` entirely.

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 3]
```diff
@@ -31,9 +33,10 @@ case class Keep(
     sourceAttributionId: Option[Id[KeepSourceAttribution]] = None,
     note: Option[String] = None,
     originalKeeperId: Option[Id[User]] = None,
-    organizationId: Option[Id[Organization]] = None) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
+    organizationId: Option[Id[Organization]] = None,
+    entitiesHash: Option[EntitiesHash] = None) extends ModelWithExternalId[Keep] with ModelWithState[Keep] with ModelWithSeqNumber[Keep] {
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: `Option` is temporary until the data is backfilled, I assume?

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: Correct.

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 4]
```diff
@@ -68,8 +71,14 @@ case class Keep(
     organizationId = lib.organizationId
   )
 
+  def withEntities(libraries: Set[Id[Library]], users: Set[Id[User]]): Keep = this.copy(entitiesHash = Some(KeepConnectionEntities(libraries, users).hash))
+
   def isActive: Boolean = state == KeepStates.ACTIVE && isPrimary // isPrimary will be removed shortly
   def isInactive: Boolean = state == KeepStates.INACTIVE
+
+  def canBeMergedInto(other: Keep): Boolean = {
+    entitiesHash == other.entitiesHash && keptAt > other.keptAt
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: `None == None` might be a problem here. As the db stands right now with this, everything can be merged into the very first keep. Probably not intended?

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: Hrm. So this method is probably named kind of poorly.

In KeepCommander, before any "merging" is done, I pre-check that the actual `KeepConnectionEntities` match up, which necessarily means that the `entitiesHash` must match. I double-check it here, just to be sure, but really this method is supposed to be "given that these two keeps have exactly the same entities connected to them, and that they have the same URI, should we merge them?"

Right now, the answer is always "yes", and we always merge the older keep into the newer keep (by "merge" I mean we deactivate the old keep). Going forward, the answer should be "as long as the metadata on the keeps is compatible (i.e., the same title, same note, and no discussion).

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: I'll probably rename this method to `hasStrictlyLessMetadata` or something along those lines. That will make it clear that it is insufficient to check only that before merging.

Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala]: :+1:

**Pull request merged**